src/reactors/NonLinearDutchOrderReactor.sol

```diff
@@ -0,0 +1,141 @@
+    {
+        NonLinearDutchOrder memory order = abi.decode(signedOrder.order, (NonLinearDutchOrder));
+        // hash the order _before_ overriding amounts, as this is the hash the user would have signed
+        bytes32 orderHash = order.hash();
+
+        if (order.info.deadline < block.timestamp) {
+            revert DeadlineReached();
+        }
+        _validateOrder(orderHash, order);
+        _updateWithCosignerAmounts(order);
+
+        resolvedOrder = ResolvedOrder({
+            info: order.info,
+            input: order.baseInput.decay(order.cosignerData.decayStartBlock),
+            outputs: order.baseOutputs.decay(order.cosignerData.decayStartBlock),
+            sig: signedOrder.sig,
+            hash: orderHash
+        });
+        resolvedOrder.handleExclusiveOverride(
+            order.cosignerData.exclusiveFiller,
+            order.cosignerData.decayStartBlock,
+            order.cosignerData.exclusivityOverrideBps,
+            false
+        );
+    }
+
+    /// @inheritdoc BaseReactor
+    function _transferInputTokens(ResolvedOrder memory order, address to) internal override {
+        permit2.permitWitnessTransferFrom(
+            order.toPermit(),
+            order.transferDetails(to),
+            order.info.swapper,
+            order.hash,
+            NonLinearDutchOrderLib.PERMIT2_ORDER_TYPE,
+            order.sig
+        );
+    }
+
+    function _updateWithCosignerAmounts(NonLinearDutchOrder memory order) internal pure {
+        if (order.cosignerData.inputAmount != 0) {
+            if (order.cosignerData.inputAmount > order.baseInput.startAmount) {
+                revert InvalidCosignerInput();
+            }
+            order.baseInput.startAmount = order.cosignerData.inputAmount;
+        }
+
+        if (order.cosignerData.outputAmounts.length != order.baseOutputs.length) {
+            revert InvalidCosignerOutput();
+        }
+        for (uint256 i = 0; i < order.baseOutputs.length; i++) {
+            NonLinearDutchOutput memory output = order.baseOutputs[i];
+            uint256 outputAmount = order.cosignerData.outputAmounts[i];
+            if (outputAmount != 0) {
+                if (outputAmount < output.startAmount) {
+                    revert InvalidCosignerOutput();
+                }
+                output.startAmount = outputAmount;
+            }
+        }
+    }
+
+    /// @notice validate the dutch order fields
+    /// - decay curves are defined
+    /// - deadline must have not passed
+    /// @dev Throws if the order is invalid
+    function _validateOrder(bytes32 orderHash, NonLinearDutchOrder memory order) internal pure {
+        if (order.baseInput.curve.relativeAmount.length == 0 || order.baseInput.curve.relativeAmount.length > 16) {
+            revert InvalidDecayCurve();
+        }
+        for (uint256 i = 0; i < order.baseOutputs.length; i++) {
+            if (order.baseOutputs[i].curve.relativeAmount.length == 0) {
```

Comment:
also this check is inconsistent with the above one which checks the max length of relativeAmount

Reply:
Just moved these to `src/lib/NonLinearDutchDecayLib.sol` to save on gas (we're looping through outputs there already).